### ISSUE TITLE
Transplant DOM node reflectors when adopting nodes between globals

### ISSUE THREAD

**Issue by jdm** [servo/servo#16515]
```html
<html>
  <body>
    <iframe id="child" src="./adopt-then-set-domain-child.html"></iframe>
  </body>
  <script>
    var child = document.getElementById("child");
    child.onload = function() {
      var node1 = child.contentDocument.getElementById("node1");
      var node2 = child.contentDocument.getElementById("node2");
      document.body.appendChild(node1);
      console.log("Child is same-origin-domain: " + (child.contentDocument !== null));
      document.domain = document.domain;
      console.log("Child is now not same-origin-domain: " + (child.contentDocument === null));
      console.log("No security exception on accessing adopted node: " + node1.innerHTML);
      try {
        console.log(node2.innerHTML);
      } catch (exn) {
        console.log("Security exception on accessing non-adopted node.");
      }
    };
  </script>
</html>
```
```html
<html>
  <body>
    <p id="node1">Hello world.</p>
    <p id="node2">Hello world.</p>
  </body>
</html>
```

**Comment by jdm** [comment]
In gecko, https://dxr.mozilla.org/mozilla-central/rev/a374c35469935a874fefe64d3e07003fc5bc8884/dom/base/nsDocument.cpp#7674-7675 ends up invoking https://dxr.mozilla.org/mozilla-central/rev/a374c35469935a874fefe64d3e07003fc5bc8884/dom/base/nsNodeUtils.cpp#577-583. [ReparentWrapper](https://dxr.mozilla.org/mozilla-central/rev/a374c35469935a874fefe64d3e07003fc5bc8884/dom/bindings/BindingUtils.cpp#2114) does the actual transplant through xpc::TransplantObject.

**Comment by asajeffrey** [comment]
IRC conversation: http://logs.glob.uno/?c=mozilla%23servo&s=18+Apr+2017&e=18+Apr+2017#c652467

TL;DR: adopting a node requires brain-transplanting it and all its descendants.

cc @avadacatavra

**Comment by asajeffrey** [comment]
#16520 is a test case for this issue.

**Comment by jdm** [comment]
As of https://bugzilla.mozilla.org/show_bug.cgi?id=1470017 I'm not actually sure if this is necessary any more. We just may need to be more deliberate about the difference between `.global()` and `.owner_global()`.
